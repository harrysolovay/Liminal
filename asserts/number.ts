import { assert } from "../util/mod.ts"

export function min(value: number, min: number) {
  assert(value >= min, `Must be gte ${min}; received ${value}.`)
}

export function max(value: number, max: number) {
  assert(value <= max, `Must be lte ${max}; received ${value}.`)
}
