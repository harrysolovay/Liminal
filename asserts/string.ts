import { assert } from "../util/mod.ts"

export function minLength(value: string, minLength: number) {
  assert(value.length >= minLength, `Length must be gte ${minLength} chars.`)
}

export function maxLength(value: string, maxLength: number) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} chars.`)
}
