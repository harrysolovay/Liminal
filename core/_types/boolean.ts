import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export const boolean: Type<boolean> = declareType(() => boolean)
