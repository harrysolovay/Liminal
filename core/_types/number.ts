import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export const number: Type<number> = declareType(() => number)
