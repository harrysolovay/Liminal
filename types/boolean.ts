import { declare, type Type } from "../core/mod.ts"

export const boolean: Type.Initial<boolean> = declare({
  name: "boolean",
  source: {
    getType: () => boolean,
  },
  subschema: () => ({ type: "boolean" }),
})
