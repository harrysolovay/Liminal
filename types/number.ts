import { declareType, type Type } from "../core/mod.ts"

export const number: Type<number> = declareType({
  name: "number",
  source: {
    getType: () => number,
  },
})
