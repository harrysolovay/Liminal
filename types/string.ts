import { declare, type Type } from "../core/mod.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare({
  name: "string",
  source: {
    getType: () => string,
  },
  refinements: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      "`minLength` cannot be greater than `maxLength`.",
    )
    return {
      minLength: `Length must be gte ${minLength} chars.`,
      maxLength: `Length must be lte ${maxLength} chars.`,
    }
  },
  subschema: () => ({ type: "string" }),
  output: (f) =>
    f({
      refinementPredicates: {
        minLength: (value, minLength) => value.length >= minLength,
        maxLength: (value, maxLength) => value.length <= maxLength,
      },
    }),
})
