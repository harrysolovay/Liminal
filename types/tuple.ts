import { declare, type Type } from "../core/mod.ts"

export function tuple<E extends Array<Type>>(
  ...elements: E
): Type<{ [K in keyof E]: E[K]["T"] }, {}, E[number]["P"]> {
  const { length } = elements
  return declare({
    name: "tuple",
    source: {
      factory: tuple,
      args: { elements },
    },
    subschema: (visit) => ({
      type: "object",
      properties: Object.fromEntries(Array.from({ length }, (_0, i) => [i, visit(elements[i]!)])),
      required: Array.from({ length }, (_0, i) => i.toString()),
      additionalProperties: false,
    }),
    output: (f) =>
      f<{ [K in keyof E]: unknown }>({
        visitor: (value, ctx) =>
          Array.from({ length }, (_0, i) =>
            ctx.visit({
              value: value[i],
              type: elements[i]!,
              junctions: {
                value: i,
                type: "number",
              },
            })) as never,
      }),
  })
}
