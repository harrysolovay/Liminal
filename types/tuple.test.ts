import { assertTypeSnapshot } from "testing"
import { boolean } from "./boolean.ts"
import { number } from "./number.ts"
import { string } from "./string.ts"
import { tuple } from "./tuple.ts"

Deno.test("tuple", async (t) => {
  await assertTypeSnapshot(t, tuple(boolean, number, string))
})
