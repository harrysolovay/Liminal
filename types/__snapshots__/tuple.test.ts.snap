export const snapshot = {};

snapshot[`tuple 1`] = `
{
  additionalProperties: false,
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;

snapshot[`tuple 2`] = `
{
  additionalProperties: false,
  description: "One.",
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;

snapshot[`tuple 3`] = `
{
  additionalProperties: false,
  description: "One. Two.",
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;
